Use services for determining mime type and filename

awead opened the pull request:
This will ultimately relate to fixing projecthydra/sufia#2553. We interrogate an incoming object about its original name and mime type, providing a values ffor each based on some logic. Testing this was difficult when it was buried in a private method. This surfaces the methods as services which can be more easily tested, as well as exposed to other consuming applications should they wish to make use of them.

mjgiarlo commented on lib/hydra/works/services/determine_mime_type.rb:
```diff
@@ -0,0 +1,24 @@
+    # @param [IO, File, Rack::Multipart::UploadedFile, #read] file
+    # @param [String, NilClass] original_name of the file
+    # @return [String]
+    class << self
```
I don't feel strongly about this, especially given we use different patterns in the hydra-works codebase, but I have a slight preference for the service object design that uses a class method to create an instance. Like so:

https://github.com/projecthydra/hydra-works/blob/master/lib/hydra/works/services/characterization_service.rb#L9-L21

That way you can make use of `attr_accessor` instead of passing around the same arguments (e.g., `file` and `original_name`). Like I said, slight preference.

awead replied:
@mjgiarlo I was going by what's currently in the repo. The majority use `self.call` without an additional initializer, but I share your opinions. The primary motivation behind doing this was to avoid having to use Rspec's `allow_any_instance_of` because the original `determine_mime_type` method was nested deep in a private method, in an internal class. I  found it too difficult to mock that properly, and it highlighted the fact that `AddFileToFileSet` was too concerned with mime types and names.

True, it is basically a module, but in a service interface it's more accessible to another application. Of course this is a bit YAGNI, but like I said, the primary motivation for this was to get around mocking issues and adhering more to single-responsibility principles. I could have gone with the module design, but a service, I think, is an extra bonus in this case.

mjgiarlo replied:
:ok_hand:

awead replied:
FWIW, I would love to see some agreement on how we write our services and have some kind of automated enforcement that ensures we're consistent. Ruboservice anyone?

mjgiarlo replied:
At the very minimum, it might be nice to arrive at two patterns:

* a Module-like pattern (a bunch of methods defined on `self`)
* a Class-like pattern (a single method defined on `self` named `#run` or `#call` that instantiates the class, uses attrs to encapsulate state, and assigns passed in params to ivars in `#initialize`) -- and it'd be great to standardize on either `#run` or `#call`

mjgiarlo commented on lib/hydra/works/services/determine_mime_type.rb:
```diff
@@ -0,0 +1,24 @@
+module Hydra::Works
+  class DetermineMimeType
+    # Determines the mime type for a given file
```
I believe this method documentation should sit above the `#call` method definition rather than above `class << self`.

mjgiarlo commented on lib/hydra/works/services/determine_original_name.rb:
```diff
@@ -0,0 +1,15 @@
+    # Determines the original name for a given file
+    # @param [IO, File, Rack::Multipart::UploadedFile, #read] file
+    # @return [String]
+    class << self
```
Same comments as above re: service object design and moving the method docs down a line.

mjgiarlo commented on lib/hydra/works/services/determine_mime_type.rb:
```diff
@@ -0,0 +1,24 @@
+      end
+
+      def mime_type_from_name_or_path(file, original_name)
+        return Hydra::PCDM::GetMimeTypeForFile.call(original_name) if original_name.present?
```
This looks like a new condition (not seeing this above). Is this a bug fix change in addition to the refactor?

awead replied:
@mjgiarlo it's going to relate to a forthcoming PR in projecthydra/curation_concerns@3cf65ca3352fad35d75e0f8505e623d4b45a0a6f that ultimately gets the original Sufia issue to pass. Do you see where I am here? Three layers down, and three gem releases just to fix one issue. 😉 

There are cases when we want to find the mime type using the provided original name, and _not_ the path. Examples are when my file has been uploaded as `sample.txt1234sdfasdf234`. The path doesn't help me there, so if the file is always represented as an object that responds to `original_name` or `original_filename`, I can supply the right name as opposed to having to rely on the path.

mjgiarlo replied:
OK! I figured. :)

mjgiarlo commented on spec/hydra/works/services/add_file_to_file_set_spec.rb:
```diff
@@ -67,14 +67,6 @@
     end
   end
 
-  context 'file responds to :path but not to :mime_type nor :original_name' do
```
Is it worth changing this test to just make sure both of the new service objects are called (but otherwise defer testing said objects to the new specs below)?

awead replied:
Seems reasonable. It integrates the service with the other two.

awead commented:
@mjgiarlo ok, fixed it up. Incidentally, I changed `DetermineMimeType` to use the call/initializer strategy. Upon doing that, rubocop started complaining that `Assignment Branch Condition size for determine_mime_type is too high` 😠  Necessitating private methods.

mjgiarlo commented:
@awead RUBOCOP. No good deed, eh? :clap:

awead commented:
thanks, @mjgiarlo !
